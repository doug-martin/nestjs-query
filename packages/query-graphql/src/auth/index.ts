export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
