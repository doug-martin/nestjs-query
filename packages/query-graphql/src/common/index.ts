export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
