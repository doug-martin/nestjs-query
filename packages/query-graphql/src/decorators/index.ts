export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
