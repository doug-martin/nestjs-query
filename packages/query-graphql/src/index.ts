export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
