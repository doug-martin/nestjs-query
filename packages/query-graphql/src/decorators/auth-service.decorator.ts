import { Class, MetaValue, ValueReflector } from '@nestjs-query/core';
import { CRUDAuthService } from '../auth';
import { AUTH_SERVICE_KEY } from './constants';

const reflector = new ValueReflector(AUTH_SERVICE_KEY);
export function AuthorizationService<DTO>(DTOClass: Class<DTO>) {
  return (AuthService: Class<CRUDAuthService<DTO>>) => {
    reflector.set(DTOClass, AuthService);
  };
}

export const getAuthService = <DTO>(DTOClass: Class<DTO>): MetaValue<Class<CRUDAuthService<DTO>>> => {
  return reflector.get(DTOClass);
};
