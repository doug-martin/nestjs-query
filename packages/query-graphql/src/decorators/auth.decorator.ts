import { Class, MetaValue, classMetadataDecorator, getClassMetadata } from '@nestjs-query/core';
import { AUTH_KEY } from './constants';
import { CRUDAuthOptions } from '../auth';

export const CRUDAuth = classMetadataDecorator<CRUDAuthOptions<any>>(AUTH_KEY);

export const getCRUDAuth = <DTO>(Cls: Class<DTO>): MetaValue<CRUDAuthOptions<DTO>> => {
  return getClassMetadata(Cls, AUTH_KEY);
};
