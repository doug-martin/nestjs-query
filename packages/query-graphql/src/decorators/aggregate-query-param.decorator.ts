import { AggregateQuery } from '@nestjs-query/core';
import { GraphQLResolveInfo } from 'graphql';
import { GqlExecutionContext } from '@nestjs/graphql';
import { createParamDecorator, ExecutionContext } from '@nestjs/common';
import graphqlFields from 'graphql-fields';

const EXCLUDED_FIELDS = ['__typename'];
const QUERY_OPERATORS: (keyof AggregateQuery<unknown>)[] = ['groupBy', 'count', 'avg', 'sum', 'min', 'max'];
export const AggregateQueryParam = createParamDecorator(<DTO>(data: unknown, ctx: ExecutionContext) => {
  const info = GqlExecutionContext.create(ctx).getInfo<GraphQLResolveInfo>();
  const fields = graphqlFields(info, {}, { excludedFields: EXCLUDED_FIELDS }) as Record<
    keyof AggregateQuery<DTO>,
    Record<keyof DTO, unknown>
  >;
  return QUERY_OPERATORS.filter((operator) => !!fields[operator]).reduce((query, operator) => {
    const queryFields = Object.keys(fields[operator]) as (keyof DTO)[];
    if (queryFields && queryFields.length) {
      return { ...query, [operator]: queryFields };
    }
    return query;
  }, {} as AggregateQuery<DTO>);
});
