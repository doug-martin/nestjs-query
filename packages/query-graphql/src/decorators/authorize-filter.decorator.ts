import { ModifyRelationOptions } from '@nestjs-query/core';
import { createParamDecorator, ExecutionContext } from '@nestjs/common';
import { GqlExecutionContext } from '@nestjs/graphql';
import { AuthorizationContext, OperationGroup } from '../auth';
import { AuthorizerContext } from '../interceptors';

type PartialAuthorizationContext = Partial<AuthorizationContext> &
  Pick<AuthorizationContext, 'operationGroup' | 'many'>;

function getContext<C>(executionContext: ExecutionContext): C {
  const gqlExecutionContext = GqlExecutionContext.create(executionContext);
  return gqlExecutionContext.getContext<C>();
}

function getAuthorizerFilter<C extends AuthorizerContext<unknown>>(
  context: C,
  authorizationContext: AuthorizationContext,
) {
  if (!context.authorizer) {
    return undefined;
  }
  return context.authorizer.authorize(context, authorizationContext);
}

function getRelationAuthFilter<C extends AuthorizerContext<unknown>>(
  context: C,
  relationName: string,
  authorizationContext: AuthorizationContext,
) {
  if (!context.authorizer) {
    return undefined;
  }
  return context.authorizer.authorizeRelation(relationName, context, authorizationContext);
}

function getAuthorizationContext(
  methodName: string | symbol,
  partialAuthContext?: PartialAuthorizationContext,
): AuthorizationContext {
  if (!partialAuthContext)
    return new Proxy<AuthorizationContext>({} as AuthorizationContext, {
      get: () => {
        throw new Error(
          `No AuthorizationContext available for method ${methodName.toString()}! Make sure that you provide an AuthorizationContext to your custom methods as argument of the @AuthorizerFilter decorator.`,
        );
      },
    });

  return {
    operationName: methodName.toString(),
    readonly:
      partialAuthContext.operationGroup === OperationGroup.READ ||
      partialAuthContext.operationGroup === OperationGroup.AGGREGATE,
    ...partialAuthContext,
  };
}

export function AuthorizerFilter<DTO>(partialAuthContext?: PartialAuthorizationContext): ParameterDecorator {
  // eslint-disable-next-line @typescript-eslint/ban-types
  return (target: Object, propertyKey: string | symbol, parameterIndex: number) => {
    const authorizationContext = getAuthorizationContext(propertyKey, partialAuthContext);
    return createParamDecorator((data: unknown, executionContext: ExecutionContext) =>
      getAuthorizerFilter(getContext<AuthorizerContext<DTO>>(executionContext), authorizationContext),
    )()(target, propertyKey, parameterIndex);
  };
}

export function RelationAuthorizerFilter<DTO>(
  relationName: string,
  partialAuthContext?: PartialAuthorizationContext,
): ParameterDecorator {
  // eslint-disable-next-line @typescript-eslint/ban-types
  return (target: Object, propertyKey: string | symbol, parameterIndex: number) => {
    const authorizationContext = getAuthorizationContext(propertyKey, partialAuthContext);
    return createParamDecorator((data: unknown, executionContext: ExecutionContext) =>
      getRelationAuthFilter(getContext<AuthorizerContext<DTO>>(executionContext), relationName, authorizationContext),
    )()(target, propertyKey, parameterIndex);
  };
}

export function ModifyRelationAuthorizerFilter<DTO>(
  relationName: string,
  partialAuthContext?: PartialAuthorizationContext,
): ParameterDecorator {
  // eslint-disable-next-line @typescript-eslint/ban-types
  return (target: Object, propertyKey: string | symbol, parameterIndex: number) => {
    const authorizationContext = getAuthorizationContext(propertyKey, partialAuthContext);
    return createParamDecorator(
      async (data: unknown, executionContext: ExecutionContext): Promise<ModifyRelationOptions<unknown, unknown>> => {
        const context = getContext<AuthorizerContext<DTO>>(executionContext);
        return {
          filter: await getAuthorizerFilter(context, authorizationContext),
          relationFilter: await getRelationAuthFilter(context, relationName, authorizationContext),
        };
      },
    )()(target, propertyKey, parameterIndex);
  };
}
