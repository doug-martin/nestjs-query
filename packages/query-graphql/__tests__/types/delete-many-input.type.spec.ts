import { plainToClass } from 'class-transformer';
import { validateSync } from 'class-validator';
import { InputType, Resolver, Query, Args, Int, ObjectType } from '@nestjs/graphql';
import { DeleteManyInputType } from '../../src';
import { FilterableField } from '../../src/decorators';
import { generateSchema } from '../__fixtures__';

describe('DeleteManyInputType', (): void => {
  @ObjectType()
  class DeleteManyDTO {
    @FilterableField()
    field!: string;
  }

  @InputType()
  class DeleteMany extends DeleteManyInputType(DeleteManyDTO) {}

  it('should create an args type with the field as the type', async () => {
    @Resolver()
    class DeleteManyInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: DeleteMany): number {
        return 1;
      }
    }
    const schema = await generateSchema([DeleteManyInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  describe('validation', () => {
    it('should validate the filter is defined', () => {
      const input = {};
      const it = plainToClass(DeleteMany, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmptyObject: 'filter must be a non-empty object',
          },
          property: 'filter',
          target: input,
        },
      ]);
    });

    it('should validate the filter is not empty', () => {
      const input = { filter: {} };
      const it = plainToClass(DeleteMany, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmptyObject: 'filter must be a non-empty object',
          },
          property: 'filter',
          target: input,
          value: input.filter,
        },
      ]);
    });
  });
});
