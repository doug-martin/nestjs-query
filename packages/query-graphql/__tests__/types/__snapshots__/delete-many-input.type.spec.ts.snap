// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteManyInputType should create an args type with the field as the type 1`] = `
type Query {
  test(input: DeleteMany!): Int!
}

input DeleteMany {
  """Filter to find records to delete"""
  filter: DeleteManyDTODeleteFilter!
}

input DeleteManyDTODeleteFilter {
  and: [DeleteManyDTODeleteFilter!]
  or: [DeleteManyDTODeleteFilter!]
  field: StringFieldComparison
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

`;
